consensus/src/dag/dag_driver.rs

```diff
@@ -134,20 +133,31 @@ impl DagDriver {
                 .prefetch_payload_data(node.payload(), node.metadata().timestamp());
             dag_writer.add_node(node)?;
 
-            let highest_round = dag_writer.highest_round();
-            dag_writer
-                .get_strong_links_for_round(highest_round, &self.epoch_state.verifier)
-                .map_or_else(|| highest_round.saturating_sub(1), |_| highest_round)
+            let highest_strong_links_round =
+                dag_writer.highest_strong_links_round(&self.epoch_state.verifier);
+            (
+                highest_strong_links_round,
+                // unwrap is for round 0
+                dag_writer
+                    .get_strong_links_for_round(
+                        highest_strong_links_round,
+                        &self.epoch_state.verifier,
+                    )
+                    .unwrap_or(vec![]),
+            )
         };
-
-        if self.current_round <= highest_strong_links_round {
-            self.enter_new_round(highest_strong_links_round + 1).await;
-        }
+        self.round_state
+            .check_for_new_round(highest_strong_link_round, strong_links)
+            .await;
         Ok(())
     }
 
     pub async fn enter_new_round(&mut self, new_round: Round) {
+        if self.round_state.current_round() >= new_round {
+            return;
+        }
```

Comment:
Maybe assert!()?

Reply:
I was worried about duplicate event or out of order event